src/Sema.zig

```diff
@@ -9968,6 +10099,13 @@ fn coerceInMemoryAllowed(dest_type: Type, src_type: Type, dest_is_mut: bool) InM
             return .no_match;
         }
 
+        // TODO NO_COMMIT understand this assert.
+        if (dest_info.@"align" == 0 and
+            src_info.@"align" == 0)
+        {
+            return .ok;
+        }
```

Comment:
The asserts were there because the logic below which does a comparison was written with the (incorrect) assumption that there would never be a 0 align value. In other words I was focused on a different change, and so was sloppy here. The logic needs to be improved to handle this case.

A value of 0 means to use the ABI alignment of the pointee type. If either of the values are zero, it means we have to ask for the abi alignment of the pointee types of both of them, an action which may cause a "dependency loop detected" compile error. For now I suggest this code:

```zig
        // If both pointers have alignment 0, it means they both want ABI alignment.
        // In this case, if they share the same child type, no need to resolve
        // pointee type alignment. Otherwise both pointee types must have their alignment
        // resolved and we compare the alignment numerically.
        if (src_info.@"align" != 0 or dest_info.@"align" != 0 or
            !dest_info.pointee_type.eql(src_info.pointee_type))
        {
            const src_align = src_type.ptrAlignment(target);
            const dest_align = dest_type.ptrAlignment(target);

            if (dest_align > src_align) {
                return .no_match;
            }
        }
```

At some point we will need to go audit all the `ptrAlignment` callsites. Obtaining `target` in this function left as an exercise for the reader :stuck_out_tongue: